fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +716,207 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+                return isAlwaysTrue(compound.getChild(0)) && isAlwaysTrue(compound.getChild(1));
+            } else if (compound.isOr()) {
+                return isAlwaysTrue(compound.getChild(0)) || isAlwaysTrue(compound.getChild(1));
+            } else if (compound.isNot()) {
+                return isAlwaysFalse(predicate.getChild(0));
+            }
+        }
+        return false;
+    }
+
+    public static ScalarOperator splitOr(ScalarOperator src, ScalarOperator target) {
+        List<ScalarOperator> srcItems = Utils.extractDisjunctive(src);
+        List<ScalarOperator> targetItems = Utils.extractDisjunctive(target);
+        int srcLength = srcItems.size();
+        int targetLength = targetItems.size();
+        for (ScalarOperator item : srcItems) {
+            removeAll(targetItems, item);
+        }
+        if (targetItems.isEmpty() && srcLength == targetLength) {
+            // it is the same, so return true constant
+            return ConstantOperator.createBoolean(true);
+        } else if (!targetItems.isEmpty()) {
+            // the target has more or item, so return src
+            return src;
+        } else {
+            return null;
+        }
+    }
+
+    public static void removeAll(List<ScalarOperator> scalars, ScalarOperator predicate) {
+        Iterator<ScalarOperator> iter = scalars.iterator();
+        while (iter.hasNext()) {
+            ScalarOperator current = iter.next();
+            if (current.equals(predicate)) {
+                iter.remove();
+            }
+        }
+    }
+
+    public static boolean isAllEqualInnerJoin(OptExpression root) {
+        Operator operator = root.getOp();
+        if (!(operator instanceof LogicalOperator)) {
+            return false;
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            boolean isEqualPredicate = isColumnEqualPredicate(joinOperator.getOnPredicate());
+            if (joinOperator.getJoinType() != JoinOperator.INNER_JOIN || !isEqualPredicate) {
+                return false;
+            }
+        }
+        for (OptExpression child : root.getInputs()) {
+            if (!isAllEqualInnerJoin(child)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static boolean isColumnEqualPredicate(ScalarOperator predicate) {
+        if (predicate == null) {
+            return false;
+        }
+        if (predicate instanceof BinaryPredicateOperator) {
+            BinaryPredicateOperator binaryPredicate = (BinaryPredicateOperator) predicate;
+            if (binaryPredicate.getBinaryType().isEqual()
+                    && binaryPredicate.getChild(0).isColumnRef()
+                    && binaryPredicate.getChild(1).isColumnRef()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static Map<ColumnRefOperator, ScalarOperator> getProjectionMap(
```

Comment:
bad name

Reply:
change it to getColumnRefMap?

Reply:
I think maybe you need move it to LineageFactory, not rename

Reply:
If these methods are only used by MV and are not really generic, then it's best to replace them in your using class, not Utils

Reply:
it is not just used in LineageFactory. it is also used in MaterializedViewRewriter. So I put it here

Reply:
Ok, I get it, but getProjectionMap/getColumnRefMap is really confusing